Modules/_datetimemodule.c

```diff
@@ -6869,24 +6966,39 @@ create_timezone_from_delta(int days, int sec, int ms, int normalize)
 }
 
 static int
-init_state(datetime_state *st, PyTypeObject *PyDateTime_IsoCalendarDateType)
-{
-    // While datetime uses global module "state", we unly initialize it once.
-    // The PyLong objects created here (once per process) are not decref'd.
-    if (st->initialized) {
-        return 0;
-    }
-
-    /* Static types exposed by the C-API. */
-    st->date_type = &PyDateTime_DateType;
-    st->datetime_type = &PyDateTime_DateTimeType;
-    st->delta_type = &PyDateTime_DeltaType;
-    st->time_type = &PyDateTime_TimeType;
-    st->tzinfo_type = &PyDateTime_TZInfoType;
-    st->timezone_type = &PyDateTime_TimeZoneType;
+copy_state(datetime_state *st, datetime_state *from)
+{
+    *st = (datetime_state){
+        .isocalendar_date_type
+                = (PyTypeObject *)Py_NewRef(from->isocalendar_date_type),
+        .us_per_ms = Py_NewRef(from->us_per_ms),
+        .us_per_second = Py_NewRef(from->us_per_second),
+        .us_per_minute = Py_NewRef(from->us_per_minute),
+        .us_per_hour = Py_NewRef(from->us_per_hour),
+        .us_per_day = Py_NewRef(from->us_per_day),
+        .us_per_week = Py_NewRef(from->us_per_week),
+        .seconds_per_day = Py_NewRef(from->seconds_per_day),
+        .epoch = Py_NewRef(from->epoch),
+    };
+    return 0;
+}
```

Comment:
As noted above, when the module is loaded we populate the module state.  If the module was loaded previously and that old module hasn't been destroyed yet, we can simply re-use all the objects in the old state.  We only need to take a new reference for each of them.  We can do this because all of the objects are essentially immutable.

In addition to being more efficient. this also preserves the same continuity of object identity for these objects that they had before.